server/src/main/java/org/apache/druid/segment/realtime/appenderator/AppenderatorImpl.java

```diff
@@ -617,6 +625,21 @@ public Object call() throws IOException
     // NB: The rows are still in memory until they're done persisting, but we only count rows in active indexes.
     rowsCurrentlyInMemory.addAndGet(-numPersistedRows);
     bytesCurrentlyInMemory.addAndGet(-bytesPersisted);
+
+    log.info("Persisted rows[%,d] and bytes[%,d]", numPersistedRows, bytesPersisted);
+
+    // bytesCurrentlyInMemory can change while persisting due to concurrent ingestion.
+    // Hence, we use bytesInMemoryBeforePersist to determine the change of this persist
+    if (bytesInMemoryBeforePersist - bytesPersisted > maxBytesTuningConfig) {
```

Comment:
I wonder if we should cut this off before we go over the limit. The heap overhead of the persisted data is going to cause additional persists to happen ever more frequently, so this seems like it will behave in a thrashing manner as it approaches the limit (though at least there is a limit to cut it off when it reaches the point of persisting every row). I don't know what an ideal cutoff point would be.

This sort of makes me wonder if including this "resident heap" calculation into the bytes in memory persist triggers should be opt in due to the potential to change behavior, but i really would prefer to not introduce another config option...

It seems ideal for us to gracefully detect conditions that are likely to cause the task to run out of memory and fail instead of just allowing it to occur, especially when using indexers instead of middle managers, so I just want to make sure we can make it well behaved.

Reply:
I also don't know what an ideal cutoff point would be. I think the thrashing should only happens for a short duration assuming the data has "many intervals/sinks" or "many columns" characteristic as the overhead increase after each persist should increase quickly. 

I added more tests to make sure it's well behaved.

I didn't add any config/opt in since most likely anything that will fail because of this would have fail because of OOM too.